### DISCUSSION TITLE
Reading/Writing Flow files optimization

### DISCUSSION THREAD

**Discussion by Protinon** [mitmproxy/mitmproxy#5763] [Ideas]
What bought this to my attention was with my personal experience loading flow files. I had saved api requests for a video game download, which means the response data was abnormally large (GBs of data...). I noticed mitm seems to load the entire file into memory before parsing it. I understand most use-cases are not mine, but this isn't optimal.

I looked at the source and also noticed mitm uses _tnetstrings_ for storing data, is there any reason to use this over the built-in [struct](https://docs.python.org/3/library/struct.html) module?

And tying both of these together, the `io.tnetstrings` module seems to only read/write `Flow` objects but currently reads an entire _tnetstring_ while staying unaware of the underlying data structure. If concrete methods were written for `io.tnetstring`, reading `Flow` objects could be optimized to ignore certain fields.

I would be willing to make contributions here.

**Comment by mhils** [comment]
> I looked at the source and also noticed mitm uses _tnetstrings_ for storing data, is there any reason to use this over the built-in [struct](https://docs.python.org/3/library/struct.html) module?

My understanding is that `struct` does not lend itself to variable-length or variable-field datastructures very well.
We tentatively want to switch to something SQLite-based at some point, and I wouldn't want to add a different format as an intermediate step.

FWIW, one workaround for this problem is to use https://docs.mitmproxy.org/dev/concepts-options/#stream_large_bodies.

**Reply by Protinon** [reply]
> FWIW, one workaround for this problem is to use https://docs.mitmproxy.org/dev/concepts-options/#stream_large_bodies.

I still need to see the response data, so I can't use this unfortunately. 

> My understanding is that `struct` does not lend itself to variable-length or variable-field datastructures very well.

`struct` is simply a Python library to make reading/writing binary files easier. From my experience, the "variable length" issue in binary files is typically solved by preceeding an array with the length.

I always to assume binary files are read/written faster than string-based storage like the current _tnetstring_ implementation, but I don't have any testing with Mitm to backup that claim.

I can look into upgrading the storage format to utilize Sqlite3. I think I can separate header data from response data to make this more optimized, but that can be tested once(if) I get a new format created.

**Reply by mhils** [reply]
> I always to assume binary files are read/written faster than string-based storage like the current tnetstring implementation, but I don't have any testing with Mitm to backup that claim.

Ah, now I see where you are coming from. Our tnetstring implementation supports binary types, i.e. length prefix + raw bytes.

> I can look into upgrading the storage format to utilize Sqlite3.

Please search the issue tracker for preceding work. This is a substantial effort, and frankly speaking, I don't have the capacity to even review it at the moment.
